### PR TITLE
New Modelica state machine prototype

### PR DESCRIPTION
**Pull request by bernhard-thiele**
- Added DAE.FLAT_SM and DAE.SM_COMP to capture state machine
  composition
- Most of the code for state machine instantiation is in the new
  package "InstStateMachineUtil.mo"
- Transformation from state machines to data-flow is now triggered
  within the front-end function
  DAEUtil.transformationsBeforeBackend(..)
- The state machine to data-flow transformation code is in the new
  package "StateMachineFlatten.mo"
  (stateMachineFlatten.stateMachineToDataFlow(..))
- The old back-end prototype in package "StateMachineFeatures.mo"
  is now obsolete

Remaining issues:
- Not yet implemented operators "ticksInState()" and "timeInState()"
  and MLS "17.3.6 Merging Connections to Multiple Outputs"
- Equations in states need to confirm to the same requirements as
  when-equations (component ref to be assigned to must be at the LHS)
- State machines are based on the synchronous elements features and
  in order simulate state-machines requires these to be working in OM
- As a current workaround to simulate state machines is is possible
  to uncomment code that wraps all equations in "when-equations" and
  replaces "previous(x)" by "pre(x)", however, this is only intended
  for development purposes
- The functionality provided by "StateMachineFlatten.mo" should be
  better adapted to a back-end implementation, however this will
  require to adapt the data structures used in the back-end
  (future work)

### CODE REVIEW

**Comment by OpenModelica-Hudson** [comment]
The job failed the coding style; for details, see [OpenModelica_TEST_PULL_REQUEST 2015-07-21_23-03-50](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/412/console). Remember to clean up the commit you did that breaks the coding style rather than piling a new commit on top of the old ones. This goes double for accidental addition of files that you need to remove.

**Comment by OpenModelica-Hudson** [comment]
The job failed to compile; for details, see [OpenModelica_TEST_PULL_REQUEST 2015-07-22_00-39-04](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/415/console).

**Comment by OpenModelica-Hudson** [comment]
The job failed to compile; for details, see [OpenModelica_TEST_PULL_REQUEST 2015-07-22_01-14-20](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/416/console).

**Comment by bernhard-thiele** [comment]
Hm, anybody an idea what could be the cause for the error reported by Hudson "Error: Internal error Graph.allReachableNodes failed."?

I can't reproduce the error on my machine.

**Comment by lochel** [comment]
That's probably due to wrong ``annotation(__OpenModelica_Interface=<something>);`` statements in new source files.

File: Compiler/boot/LoadCompilerSources.mos [hunk 1]
```diff
@@ -324,6 +327,7 @@ if true then /* Suppress output */
     "../Util/HashTableExpToIndexExp.mo",
     "../Util/HashTableCrToExpSourceTpl.mo",
     "../Util/HashTableCrToCrEqLst.mo",
+    "../Util/HashTableSM1.mo",
     "../Util/HashTableCrToUnit.mo",
```
Comment (lochel) [inline Compiler/boot/LoadCompilerSources.mos]: This causes the error: 
``
[/OpenModelica/OMCompiler/Compiler/boot/build/InstStateMachineUtil.interface.mo:1:1-81:25:writable] Error: Failed to get dependencies for package InstStateMachineUtil. Perhaps there is an import to a non-existing package.
``

**Comment by bernhard-thiele** [comment]
Thanks a lot for looking into it!

 I will put it somewhere else and try again.

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2015-07-22_11-16-32](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/418/console).

**Pull request closed without merging**